src/tango/quic/fd_quic.c

```diff
@@ -1240,11 +1271,167 @@ fd_quic_handle_v1_initial( fd_quic_t *               quic,
+          memcpy(tp->retry_source_connection_id,
+                 new_conn_id.conn_id,
+                 new_conn_id.sz);
+
+          fd_quic_retry_t retry_pkt;
+          retry_pkt.hdr_form = 1;
+          retry_pkt.fixed_bit = 1;
+          // .long_packet_type
+          retry_pkt.unused = 0xf;
+          retry_pkt.version = 1;
+          retry_pkt.dst_conn_id_len = pkt->long_hdr->src_conn_id_len;
+          // .dst_conn_id
+          retry_pkt.src_conn_id_len = new_conn_id.sz;
+          // .src_conn_id
+          // .retry_token
+          // .retry_integrity_tag
+
+          memcpy(retry_pkt.dst_conn_id, pkt->long_hdr->src_conn_id, pkt->long_hdr->src_conn_id_len);
+          memcpy(retry_pkt.src_conn_id, &new_conn_id.conn_id, retry_pkt.src_conn_id_len);
+
+          /* Retry token */
+          ulong now = fd_quic_now(quic);
+          int   rc  = fd_quic_retry_token_encrypt(
+              &orig_dst_conn_id,
+              now,
+              &new_conn_id,
+              pkt->ip4->saddr,
+              pkt->udp->net_sport,
+              retry_pkt.retry_token
+          );
+          if (FD_UNLIKELY(rc == FD_QUIC_FAILED)) {
+            return FD_QUIC_PARSE_FAIL;
+          }
+
+          /* Retry integrity tag */
+          fd_quic_retry_pseudo_t retry_pseudo_pkt = {
+              .odcid_len = orig_dst_conn_id.sz,
+              // .odcid
+              .hdr_form = retry_pkt.hdr_form,
+              .fixed_bit = retry_pkt.fixed_bit,
+              .long_packet_type = retry_pkt.long_packet_type,
+              .unused = retry_pkt.unused,
+              .version = retry_pkt.version,
+              .dst_conn_id_len = retry_pkt.dst_conn_id_len,
+              // .dst_conn_id
+              .src_conn_id_len = retry_pkt.src_conn_id_len,
+              // .src_conn_id
+              // .retry_token
+          };
+          // TODO can make this more efficient by directly manipulating the retry_pkt pkt
+          // bytes directly: prepending the ODCID and removing the retry_pkt integrity tag
+
+          // TODO zero-copy?
+          memcpy(&retry_pseudo_pkt.odcid, &orig_dst_conn_id.conn_id, orig_dst_conn_id.sz);
+          memcpy(&retry_pseudo_pkt.dst_conn_id, &retry_pkt.dst_conn_id, retry_pkt.dst_conn_id_len);
+          memcpy(&retry_pseudo_pkt.src_conn_id, &retry_pkt.src_conn_id, retry_pkt.src_conn_id_len);
+          memcpy(&retry_pseudo_pkt.retry_token, &retry_pkt.retry_token, FD_QUIC_RETRY_TOKEN_SZ);
+
+          ulong retry_pseudo_footprint = fd_quic_encode_footprint_retry_pseudo(&retry_pseudo_pkt);
+          uchar retry_pseudo_buf[retry_pseudo_footprint];
+          fd_quic_encode_retry_pseudo(retry_pseudo_buf, retry_pseudo_footprint, &retry_pseudo_pkt);
+          fd_quic_retry_integrity_tag_encrypt(retry_pseudo_buf, (int) retry_pseudo_footprint, retry_pkt.retry_integrity_tag);
+
+          ulong tx_buf_sz = fd_quic_encode_footprint_retry(&retry_pkt);
+          uchar tx_buf[tx_buf_sz];
```

Comment:
so this is a good point -- the encode footprint is actually being used in one of the tests

in general, the token should be variable-length (it's opaque in the RFC) -- it doesn't matter on the server-side because we decide the scheme (and length) but our client isn't really robust to different token lengths right now.

i tried implementing variable length encoding with the templates but got into memory hell so decided to punt for now. i probably should address that after at some point. if that sg, i'll leave this as variable-length?